### DISCUSSION TITLE
ANSI color output in webview?!?

### DISCUSSION THREAD

**Discussion by jens-maus** [community/community#26944] [Actions]
<p>Hello,</p>
<p>I just recognized that in one of my GitHub Action workflows which I ported over from travis no ANSI colors are shown in the Webview of the output. Is there some setting or are ANSI color escape sequences are simply stripped in GitHub Actions while with travis they are perfectly displayed in the webview?</p>
<p>In the actual job that is run I use the nodejs-based mocha test framework and in its output Mocha uses ANSI escape sequences to either show success with a green arrow or in case of a failure with a red cross and this works perfectly in the Travis webview while in the GitHub Actions job output view no ANSI colors are displayed at all.</p>
<p>So does anyone have an idea how to enable ANSI colors in the webview or is this simply a missing feature?!?</p>
<p>Thanks in advance.</p>

**Comment by mscoutermarsh** [comment]
<p>Yes, we support ANSI colors.</p>
<p>I haven’t used Mocha specifically. Many test runners check for an env var or terminal setting before they output color.</p>

**Comment by weide-zhou** [comment]
<p>Hi <a class="mention" href="/jens-maus">@jens-maus</a> ,</p>
<p>What do you mean of ‘ANSI color escape sequences’ here? Could you please show a screenshot in travis?</p>
<p>Checked on my side, with below command, the ANSI escape sequences works in bash output as expected.</p>
<pre><code>echo -e "\033[31;1;4mHello\033[0m"
</code></pre>
<p><img src="https://user-images.githubusercontent.com/52772761/181100897-deebd506-6435-44fb-a403-7ef49e0db8bf.png" alt="ANSIColor.png" data-base62-sha1="lmz7mAUgiKX07tadq2Ul0lh3HLZ" width="688" height="161"></p>
<p>Thanks.</p>

**Comment by jens-maus** [comment]
<blockquote>
<p><a class="mention" href="/weide-zhou">@weide-zhou</a> wrote:</p>
</blockquote>
<blockquote>
<p> </p>
</blockquote>
<blockquote>
<p>What do you mean of ‘ANSI color escape sequences’? Could you please show a screenshot?</p>
</blockquote>
<p>I mean the following. This is the output on travis, e.g.:</p>
<p><img src="https://user-images.githubusercontent.com/5887769/181100899-3b48d397-8757-4f89-a43e-a70eda1f81a0.png" alt="Bildschirmfoto 2020-02-14 um 08.50.32.png" data-base62-sha1="cjn0Xgo4Xh1I999PhAr03i2FiAs" width="468" height="494"></p>
<p>And the same output, but on GitHub Actions looks like:</p>
<p><img src="https://user-images.githubusercontent.com/5887769/181100900-381ae08e-bf62-480c-a855-c23cee8b0c34.png" alt="Bildschirmfoto 2020-02-14 um 08.51.06.png" data-base62-sha1="8cWtXGRYJ5LkCdhLXp5zdeAFzti" width="502" height="471"></p>
<p>So please note the color difference and that within GitHub Actions no colorized output is presented at all. That’s why I was wondering how to enable ANSI color output in GitHub Actions or if this is a missing feature.</p>

**Comment by jens-maus** [comment]
<blockquote>
<p><a class="mention" href="/mscoutermarsh">@mscoutermarsh</a> wrote:<br>
Yes, we support ANSI colors.</p>
</blockquote>
<blockquote>
<p>I haven’t used Mocha specifically. Many test runners check for an env var or terminal setting before they output color.</p>
</blockquote>
<p>Do you know which terminal settings or env variable as used to enable this is GitHub Actions? In fact, I already tried to set the TERM variable to “xterm-256color” right before starting the mocha test, but this didn’t help. So I was wondering if there might be a standard TERM/ENV variable which I need to setup in the workflow and if so, why it is not setup per default as in travis I never had to care about that and ANSI color output is the default.</p>

**Comment by jens-maus** [comment, marked as answer]
<p>Ok, after some more investigation I could solve the problem myself. I simply had to add the command-line option –colors when executing mocha (see <a href="https://github.com/jens-maus/occu-test/commit/0a57c8fadda3e7d4d8f2043c7171dc4d6a10966e" target="_blank" rel="nofollow noopener">https://github.com/jens-maus/occu-test/commit/0a57c8fadda3e7d4d8f2043c7171dc4d6a10966e</a>).</p>
<p>So somehow with the default environment of GitHub Actions mocha could not identify itself that it is running within a TTY enabled device and thus enable ANSI color output itself. See here (<a href="https://github.com/mochajs/mocha/issues/1304" target="_blank" rel="nofollow noopener">https://github.com/mochajs/mocha/issues/1304</a>) for similar discussions to force mocha to output ANSI colors on non tty devices. So the remaining question that might be there is, why isn’t the GitHub Actions environment not identified as a proper TTY device by mocha while within travis there isn’t any issue and mocha outputs all content using ANSI colors?!?</p>

**Comment by weide-zhou** [comment]
<p>Hi <a class="mention" href="/jens-maus">@jens-maus</a> ,</p>
<p>Glad to know it’s been resolved by adding the command-line option –colors<span> </span>when executing mocha.</p>
<p>It’s recommended to raise a ticket <a href="https://support.github.com/contact/feedback?contact%5Bcategory%5D=actions" rel="nofollow noopener">here</a> where github action managers will take a review. Thanks.</p>

**Comment by ScottRudiger** [comment]
<p>FYI, if you’re using <code>chalk</code>, you may also set the environment variable <code>FORCE_COLOR</code> to <code>0</code> for no color, <code>1</code> for more color, <code>2</code> for even more color, or <code>3</code> for the most colorful of colors:</p>
<pre><code>env:
  FORCE_COLOR: 3
</code></pre>
<p>For some libraries, though, this will not be enough or annoying warnings will print in the webview because it is  <strong>not technically a TTY</strong>. For example, <a href="https://www.npmjs.com/package/boxen" rel="nofollow noopener">boxen</a> needs a TTY so this repeated error message shows up:</p>
<p><img src="https://user-images.githubusercontent.com/26824724/181100907-62a03438-d3ec-4782-9f2c-1cde6abdca2b.png" alt="Screen Shot 2020-05-11 at 5.30.43 PM.png" data-base62-sha1="dVnHAeSXrBjvstoaxveuhJ5FqoS" width="688" height="226"></p>

**Comment by perlun** [comment]
<p>There does indeed seem to be a problem with GitHub Actions not working as one would expect with</p>
<pre><code>TERM=xterm
</code></pre>
<p>or</p>
<pre><code>TERM=xterm-256color
</code></pre>
<p>I’ve tried both of these to try and get some colored output from my xUnit tests, but neither one works (where they <em>do</em> work when running the tests locally, on Linux).</p>
<p>Here is one example of such a job: <a href="https://github.com/perlun/perlang/runs/690297760?check_suite_focus=true" rel="nofollow noopener">https://github.com/perlun/perlang/runs/690297760?check_suite_focus=true</a></p>
<p><span class="mention">@weide-zhouIf</span> no issue has been created for this yet, please let me know and I’ll file one. Not all tools support a FORCE_COLOR flag, unfortunately, so it would be important to get GitHub Actions working more flawlessly in this regard.</p>

**Comment by weide-zhou** [comment]
<p>Hi <a class="mention" href="/perlun">@perlun</a> ,</p>
<p>Thanks for your checking! Please raise a feedback/feature_request ticket <a href="https://support.github.com/contact/feedback?contact%5Bcategory%5D=actions" rel="nofollow noopener">here</a> where github product manager will take a review.</p>
<p>Regards.</p>

**Comment by etep** [comment]
For colorized logs in GH actions, here is what worked for me: `script -q -e -c "export TERM=xterm-256color; bash --noprofile --norc -eo pipefail {0}"`.

In slightly more detail, this solved the problem of GH actions interpreting ANSI color escapes that were produced by colorized logs inside of a python script inside of a `pytest`. This thread (and a few others) was the top search result, but its contents & solution were ultimately not as useful as the above.

Useful links:
[The first working solution I found.](https://discourse.nixos.org/t/github-actions-nix-shell-and-coloured-output/11479/12)
[Adding `pipefail` for robustness.](https://github.com/actions/runner/issues/241#issuecomment-842566950)
[GH docs](https://docs.github.com/en/actions/writing-workflows/workflow-syntax-for-github-actions#using-a-specific-shell)

Here is more context as a `yaml`  snippet.
```
<...>
jobs:
  test:
    runs-on: ubuntu-latest
    defaults:
      run:
        # How to fix ANSI color codes w/ GH actions:
        shell: 'script -q -e -c "export TERM=xterm-256color; bash --noprofile --norc -eo pipefail {0}"'
    steps: <...>
```
